### ISSUE TITLE
[RFC]: Third-party Sparse Attention Backends Support

### ISSUE THREAD

**Issue by BeatSeat** [vllm-project/vllm-omni#4661]
### Motivation.

PR #4493 introduced the `SPARSE_ATTN` plugin dispatcher for third-party sparse attention kernels in the diffusion (DiT) path. The infrastructure is in place — a callable contract `fn(q, k, v, params, is_causal, softmax_scale) -> Tensor` with three-tier param merge (static < per-forward < dynamic).

However, no official plugin packages exist yet for the sparse attention methods that have been discussed or prototyped in the vllm-omni ecosystem. This RFC surveys all applicable approaches for DiT and proposes official plugin packages in two stages.

### 1. Overview

vLLM-Omni currently supports six built-in diffusion attention backends (`FLASH_ATTN`, `TORCH_SDPA`, `SAGE_ATTN`, `SAGE_ATTN_3`, `CUDNN_ATTN`, `FLASHINFER_ATTN`) — all dense. For video diffusion, sparse attention can reduce the O(N²) compute to O(N·C), enabling longer videos and lower latency.

PR #4493 provides the plugin infrastructure. What's missing are the actual plugin packages that users can `pip install` and select by name in per-role config. This RFC proposes four such packages spanning both training-free and trainable sparse attention methods, prioritized by maturity.

---

#### Goals

- **Stage 1**: Ship official `pip install`-able plugin packages for SpargeAttn and LVSA, both of which have working demo code and published papers.
- **Stage 2**: Build plugin packages for SLA and SageSLA, which require fine-tuned checkpoints and depend on additional CUDA kernels.
- All plugins follow the PR #4493 contract: entry-point registration under `vllm_omni.sparse_attn`, single-function API, three-tier param merge.
- Each plugin is self-attention only per PR #4493 convention; cross-attention stays dense.
- Provide a selection guide so users can choose the right backend for their use case and model.

#### Non-goals

- SageAttention (`SAGE_ATTN`, `SAGE_ATTN_3`) — already a built-in backend, dense INT8 quantized only, no sparsity; not a plugin.
- LVSA integration into vLLM-Omni core — LVSA has its own RFC (#3115); this RFC covers only its plugin package.
- TurboDiffusion full model integration (rCM distillation, checkpoint format conversion) — that's a separate feature tracked in #443.

---

#### Architecture

```
DiffusionAttentionBackendEnum
├── FLASH_ATTN          (built-in, dense)
├── TORCH_SDPA          (built-in, dense)
├── SAGE_ATTN           (built-in, dense + INT8)
├── CUDNN_ATTN          (built-in, dense)
├── FLASHINFER_ATTN     (built-in, dense)
└── SPARSE_ATTN         plugin dispatcher (PR #4493)
    ├── spargeattn      ── SpargeAttn (training-free, Stage 1)
    ├── lvsa_sparse     ── LVSA (training-free, Stage 1)
    ├── sla             ── SLA (trainable, Stage 2)
    └── sagesla         ── SageSLA (trainable, Stage 2)
```

Each plugin is a standalone pip package that registers a single callable:

```python
# Plugin contract (from PR #4493)
def attn_fn(query, key, value, params: dict, is_causal: bool, softmax_scale: float) -> torch.Tensor:
    # q/k/v: BSND [batch, seq_len, num_heads, head_dim]
    # params merged from: static config < per-forward state < dynamic extra
    ...
    return out  # BSND, same layout
```

**Stage 1 — Training-free (demos exist, promote to official)**

| Method | Type | Training | Demo | Paper | Code | Parameters |
|--------|------|----------|------|-------|------|------------|
| **SpargeAttn** | Content-adaptive block-sparse (128×64), INT8 via SageAttention2 | None | [sparge demo](https://github.com/harenome/vllm-omni-plugin-demo/tree/main/sparge) | [ICML 2025](https://arxiv.org/abs/2502.18137) | [thu-ml/SpargeAttn](https://github.com/thu-ml/SpargeAttn) | `topk` (0~1, default 0.5) |
| **LVSA** | Structured geometry-aware block-sparse, SDPA / FlashInfer CSR | None | [lvsa demo](https://github.com/harenome/vllm-omni-plugin-demo/tree/main/lvsa) | [arXiv:2605.31057](https://arxiv.org/abs/2605.31057) | [gglorian/LVSA](https://github.com/gglorian/LVSA) | `reference_latent_frames` (21), `backend` (sdpa/flashinfer), geometry from per-forward state |

**Stage 2 — Trainable (new development)**

| Method | Type | Training | Demo | Paper | Code | Parameters | Dependencies |
|--------|------|----------|------|-------|------|------------|--------------|
| **SLA** | Sparse + Linear hybrid (BF16), BLKQ=64/BLKK=64 | Required (fine-tune ~1k steps) | — | [arXiv:2509.24006](https://arxiv.org/abs/2509.24006) | [thu-ml/SLA](https://github.com/thu-ml/SLA) | `topk` (0.2), `feature_map` (softmax/elu/relu), `BLKQ`, `BLKK` | `thu-ml/SLA` CUDA kernel |
| **SageSLA** | Sparse + Linear + INT8 quant, BLKQ=128/BLKK=64 fixed | Required (same SLA weights) | — | [TurboDiffusion arXiv:2512.16093](https://arxiv.org/abs/2512.16093) | [thu-ml/SLA/SageSLA](https://github.com/thu-ml/SLA/tree/main/SageSLA) | `topk` (0.15), `feature_map` (softmax/elu/relu) | `thu-ml/SLA/SageSLA` + `thu-ml/SpargeAttn` |

#### API & Interface Changes

Each plugin adds one entry in its `pyproject.toml`:

```toml
[project.entry-points."vllm_omni.sparse_attn"]
spargeattn = "sparge_vllm_omni:sparge_attn_fn"
lvsa_sparse = "lvsa_vllm_omni:lvsa_attn_fn"
sla = "sla_vllm_omni:sla_attn_fn"
sagesla = "sagesla_vllm_omni:sagesla_attn_fn"
```

No changes to vLLM-Omni core. The dispatcher (PR #4493) resolves these by name at model init.

User-facing CLI:

```bash
# Stage 1
vllm-omni serve Wan-AI/Wan2.2-T2V-A14B-Diffusers \
    --diffusion-attention-config.per_role.self.backend spargeattn \
    --diffusion-attention-config.per_role.self.extra.topk 0.3

# Stage 2
vllm-omni serve Wan-AI/Wan2.2-T2V-A14B-Diffusers \
    --diffusion-attention-config.per_role.self.backend sla \
    --diffusion-attention-config.per_role.self.extra.topk 0.2
```

#### Key Technical Decisions

- **Function-based plugin over full AttentionBackend subclass**: PR #4493 chose a function contract over a class hierarchy for minimal boilerplate. All four plugins follow this pattern.
- **Instance caching for SLA/SageSLA**: `SparseLinearAttention`/`SageSparseLinearAttention` objects are stateful (CUDA kernel launches, internal buffers). The plugin caches them by config key to avoid recreating on every forward pass.
- **LVSA reads geometry from per-forward state**: `total_latent_frames` and `patches_per_frame` are injected by the model transformer (PR #4493). This is essential for LVSA's (frame, patch) coordinate mapping. SpargeAttn and SLA/SageSLA do not require geometry.

#### Dependencies & Risks

| Risk | Impact | Mitigation |
|------|--------|------------|
| SpargeAttn CUDA kernel build failures | Stage 1 blocked | CI matrix across CUDA 12.x, pin known-good torch/CUDA versions |
| SLA CUDA kernel not yet packaged for pip | Stage 2 blocked | Vendor `thu-ml/SLA` as submodule; or work with thu-ml to publish wheels |
| SageSLA depends on both SLA + SpargeAttn kernels | Stage 2 increased build complexity | Document install order; consider a meta-package |
| SLA-fine-tuned checkpoints not in HF format | Stage 2 quality validation blocked | Investigate TurboDiffusion weight format; adapter script needed |

#### Performance

Expected speedup ranges (vs dense SDPA, rough estimates from published papers):

| Method | Speedup (vs SDPA) | VRAM savings | Quality impact |
|--------|------------------|-------------|---------------|
| SpargeAttn (topk=0.3) | ~1.3-1.8× | ~15% | Negligible |
| LVSA (2× horizon) | ~1.4× | ~40% (sparse attn) | Negligible |
| LVSA (4× horizon, FlashInfer) | ~2.3× | ~60% (sparse attn) | Quality improvement (avoids OOM/looping) |
| SLA (topk=0.2) | ~2-3× | ~20% | Low (post-fine-tune) |
| SageSLA (topk=0.15) | ~3-5× | ~40% | Low (post-fine-tune + INT8) |

Precise benchmarks will be gathered during each stage's validation phase.

---

### 4. Correctness & Testing plans

**Stage 1 (SpargeAttn, LVSA):**
- Unit: plugin entry-point resolution test (CPU-runnable, mocks entry points)
- Integration: generate 1 video per supported model (Wan2.2-14B, HunyuanVideo) with each backend, compare output shape and non-NaN
- Quality: VQeval composite score vs dense baseline at equal step count

**Stage 2 (SLA, SageSLA):**
- Unit: same entry-point resolution test
- Integration: load SLA-fine-tuned checkpoint, generate 1 video, verify output is non-degenerate
- Quality: VQeval composite score vs dense baseline at equal step count; compare SLA vs SageSLA quality delta

---

### 5. Open Questions & Discussions

1. **Plugin package naming**: should the PyPI packages be `sparge-vllm-omni` / `lvsa-vllm-omni` / etc., or `vllm-omni-backend-spargeattn` / a shorter form?
2. **SLA checkpoint distribution**: should we provide a HF-compatible SLA-fine-tuned checkpoint conversion script, or work with TurboDiffusion authors to publish native checkpoints?
3. **SageAttention overlap**: SageAttention is already a built-in backend. Should we add a note that `SAGE_ATTN` + SpargeAttn can be composed (SpargeAttn's sparse block selection runs on INT8 via SageAttention2)?
4. **Backward compat with PR #888 and #2231**: those PRs proposed alternative sparse attention paths. Now that PR #4493 provides the canonical approach, should we close #888 and #2231?

---

### 6. References

- PR #4493: [Plugin interface for third-party sparse attention for per-forward DiT](https://github.com/vllm-project/vllm-omni/pull/4493)
- RFC #3715: [Pluggable custom attention backends + per-forward state](https://github.com/vllm-project/vllm-omni/issues/3715)
- RFC #3115: [LVSA — Long Video Sparse Attention](https://github.com/vllm-project/vllm-omni/issues/3115)
- RFC #2233: [Plugin-based Sparse Attention Interface](https://github.com/vllm-project/vllm-omni/issues/2233)
- RFC #1568: [Extending attention backends](https://github.com/vllm-project/vllm-omni/issues/1568)
- Demo repo: [vllm-omni-plugin-demo](https://github.com/harenome/vllm-omni-plugin-demo)
- Design doc: `docs/design/feature/sparse_attention_backends_support.md`
- SpargeAttn: [thu-ml/SpargeAttn](https://github.com/thu-ml/SpargeAttn) | [ICML 2025](https://arxiv.org/abs/2502.18137)
- LVSA: [gglorian/LVSA](https://github.com/gglorian/LVSA) | [arXiv:2605.31057](https://arxiv.org/abs/2605.31057)
- SLA: [thu-ml/SLA](https://github.com/thu-ml/SLA) | [arXiv:2509.24006](https://arxiv.org/abs/2509.24006)
- SageSLA: [thu-ml/SLA/SageSLA](https://github.com/thu-ml/SLA/tree/main/SageSLA)
- TurboDiffusion: [thu-ml/TurboDiffusion](https://github.com/thu-ml/TurboDiffusion) | [arXiv:2512.16093](https://arxiv.org/abs/2512.16093)

### Feedback Period.

One week.

### CC List.

@hsliuustc0106 @harenome @gglorian @zzhang-fr @Gaohan123 

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.

**Comment by zzhang-fr** [comment]
Thanks for the RFC, @BeatSeat — fully supportive of bringing in new sparse-attention backends as plugins. (Context: I work on #4493/#3715 and LVSA.)

**About your open questions.**

1. *Naming* — a uniform prefix (e.g. `vllm-omni-sparse-*`) reads cleaner than the mixed forms; and worth locking the user-facing backend string early, since it's config API and renames hurt after release.
2. *SLA checkpoints* — native HF checkpoints from the TurboDiffusion (thu-ml) authors are cleanest; a project-maintained converter is the fallback if that's not feasible (at the cost of upkeep).
3. *SageAttn* — no special composition needed: any sparse-over-SageAttention variant is just another plugin under its own name (the dispatcher takes one named kernel per role). If such a combo is useful, ship it as its own backend.
4. *Close #888 / #2231* — #2231 is superseded by #4493, so close it once that merges. For #888, I'd defer to its author.

**One note on the contract.** The four packages all ride on the #4493 contract, so the main thing is keeping that clean — e.g. how per-forward geometry is published (currently a small per-model call) may still evolve. We'll follow up on that in #4493 / its design RFC #3715. Otherwise the plan looks solid — and packaging LVSA as a pip plugin is a good idea, which we'll consider on our end.

**Comment by gglorian** [comment]
Thanks for putting this together @BeatSeat — glad to see LVSA in the Stage 1 set, and the function-plugin route is a clean fit for the training-free methods. One correction first, then a few design notes.

**Repo link correction.** The references and the Stage 1 table point to `gglorian/LVSA`, which is my personal **fork**. The canonical repo is **`JiusiServe/LongVideoSparseAttention`** (the parent). @harenome's demo already depends on JiusiServe core (`lvsa @ git+https://github.com/JiusiServe/LongVideoSparseAttention.git@…`), so this also aligns the RFC with the prototype it cites. Please update both the "Code" column and the References to `JiusiServe/LongVideoSparseAttention`. The paper link (`arXiv:2605.31057`) is correct.

A few design notes:

1. **Package layout — match the demo (it has it right).** The entry-point table lists `lvsa_sparse = "lvsa_vllm_omni:lvsa_attn_fn"`, but that doesn't match the actual prototype, which registers `lvsa_sparse = "lvsa_sparse_fn:lvsa_attn"` — a **separate** package. That separation is deliberate and correct: `lvsa-vllm-omni` is our existing shipped package and it exposes a full `AttentionBackend` under `vllm_omni.general_plugins`, not a function — reusing that import path would collide. Suggest the table adopt the demo's distinct `lvsa_sparse_fn` package (and lock the user-facing backend string early, per Open Question #1).

2. **Import LVSA core, don't vendor it.** The prototype imports the math from `lvsa.sparse_attention` (good) but copies two pieces verbatim — the stateful FlashInfer runner and `build_global_kv` — because those currently live only in our shell (which can't be installed without the colliding backend from point 1). Verbatim copies drift from upstream, and that runner has subtle invariants (compact-KV tiling, plan-reuse across layers, LSE-merge). For the official plugin I'll expose those in the torch-only `lvsa` core so it can depend on a pinned `lvsa` release and import them instead of carrying copies (I'll coordinate with @harenome on this — filing a separate issue on the demo).

3. **Scope of the function route — two separate things:**
   - *Single-GPU* (the demo's README note) is fine as demo scope. Just worth being explicit that the official plugin should keep **context-parallel**, which our shipped backend already supports.
   - *Cosmos / dual-stream* is a **contract-level** question, not a scoping one. LVSA's Cosmos path concatenates the und/text K/V into the self-attention keys and splits it back out as a separate dense term (LSE-merged). A bare `fn(q,k,v,…)→out` has nowhere to express that — the demo sidesteps it by asserting `text == 0`. If Cosmos is in scope for this route, #4493 needs a way to carry the und split (e.g. via `params`); otherwise it's worth stating Cosmos is out of scope for the function plugin.

4. **Geometry contract stability.** LVSA is the only one of the four plugins that needs geometry (`total_latent_frames`, `patches_per_frame`, step index). @zzhang-fr already noted the per-forward geometry-publishing API may still evolve — since any change there breaks LVSA specifically, I'd like it spec'd/pinned in #4493 before the LVSA plugin ships.

5. **A coherent story across the tracks.** Between our shipped `general_plugins` backend (CP + Cosmos), this `sparse_attn` function plugin, and the core-integration RFC #3115, there are three overlapping LVSA paths. It'd help users to state the intended end-state and which is recommended, rather than leave three options.

Minor: the `reference_latent_frames` default `21` is Wan-specific (33 for HunyuanVideo, 48 for Cosmos) — the example/default should make clear it's set per model, or sparsity is mis-tuned.

Glad to help with the LVSA packaging once the contract above is settled.
